--- bootstrap	2010-06-28 17:15:56.000000000 +0200
+++ bootstrap-patched	2010-09-18 00:24:29.000000000 +0200
@@ -1426,7 +1426,7 @@
 export MAKE

 # Run bootstrap CMake to configure real CMake
-"${cmake_bootstrap_dir}/cmake" "${cmake_source_dir}" "-C${cmake_bootstrap_dir}/InitialCacheFlags.cmake" "-G${cmake_bootstrap_generator}" ${cmake_bootstrap_system_libs}
+"${cmake_bootstrap_dir}/cmake" "${cmake_source_dir}" "-C${cmake_bootstrap_dir}/InitialCacheFlags.cmake" "-G${cmake_bootstrap_generator}" -DCMAKE_BUILD_TYPE=Debug ${cmake_bootstrap_system_libs}
 RES=$?
 if [ "${RES}" -ne "0" ]; then
   cmake_error 11 "Problem while running initial CMake"
